src/adc.rs

```diff
@@ -449,6 +449,244 @@ impl_adc!(ADC1: adc_unit_t_ADC_UNIT_1);
+                    atten,
+                    bitwidth,
+                    ..Default::default()
+                };
+                let mut cal_handle: adc_cali_handle_t = core::ptr::null_mut();
+                if let Err(_err) = unsafe {
+                    esp!(esp_idf_sys::adc_cali_create_scheme_line_fitting(
+                        &cal_config,
+                        &mut cal_handle
+                    ))
+                } {
+                    // I'd log a warning but the log crate is not available here
+                    None
+                } else {
+                    Some(cal_handle)
+                }
+            }
+            #[cfg(not(any(esp32, esp32c2, esp32s2)))]
+            None
+        }
+
+        fn pin(&mut self) -> &mut PeripheralRef<'d, T> {
+            &mut self.pin
+        }
+    }
+
+    pub struct AdcDriver<'d> {
+        handle: adc_oneshot_unit_handle_t,
+        adc: u8,
+        _ref: PhantomData<&'d ()>,
+    }
+
+    impl<'d> AdcDriver<'d> {
+        pub fn new<A: Adc>(_adc: impl Peripheral<P = A> + 'd) -> Result<Self, EspError> {
+            let config = adc_oneshot_unit_init_cfg_t {
+                unit_id: A::unit(),
+                ..Default::default()
+            };
+            let mut adc_handle: adc_oneshot_unit_handle_t = core::ptr::null_mut();
+            unsafe { esp!(adc_oneshot_new_unit(&config, &mut adc_handle))? };
+            Ok(Self {
+                handle: adc_handle,
+                adc: A::unit() as u8,
+                _ref: PhantomData,
+            })
+        }
+
+        pub fn read<T>(&mut self, channel: &mut AdcChannelDriver<'_, T>) -> Result<u16, EspError>
+        where
+            T: ADCPin,
+        {
+            let raw = self.read_raw(channel)?;
+            self.raw_to_cal(raw, channel)
+        }
+
+        #[inline(always)]
+        pub fn read_raw<T>(
+            &mut self,
+            channel: &mut AdcChannelDriver<'_, T>,
+        ) -> Result<u16, EspError>
+        where
+            T: ADCPin,
+        {
+            let mut measurement = 0;
+            unsafe {
+                esp!(adc_oneshot_read(
+                    self.handle,
+                    channel.pin().adc_channel(),
+                    &mut measurement
+                ))
+            }?;
+            Ok(measurement as u16)
+        }
+
+        #[inline(always)]
+        pub fn raw_to_cal<T>(
+            &mut self,
+            raw: u16,
+            channel: &AdcChannelDriver<'_, T>,
+        ) -> Result<u16, EspError>
+        where
+            T: ADCPin,
+        {
+            let value = if let Some(calibration) = &channel.calibration {
+                let mut mv = 0i32;
+                unsafe {
+                    esp!(adc_cali_raw_to_voltage(*calibration, raw as i32, &mut mv))?;
+                };
+                mv as u16
+            } else {
+                raw
+            };
+            Ok(value)
+        }
+    }
+}
+
```

Comment:
Needs impl of the e-hal 0.2 trait. This trait is terrible, yet the only thing we have, so good to have it implemented, just like it is for the 4.X driver.

Reply:
OK.  I'm struggling with this one.  Maybe I'm way off base but I'm attempting to implement something like this 
```rust
    impl<'d, T, M, PIN> embedded_hal_0_2::adc::OneShot<T::Adc, u16, T> for AdcDriver<'d, T::Adc>
    where
        T: ADCPin,
        M: Borrow<AdcDriver<'d, T::Adc>>,
        PIN: embedded_hal_0_2::adc::Channel<T::Adc, ID = AdcChannelDriver<'d, T, M>>,
    {
        type Error = EspError;
    
        fn read(&mut self, _pin: &mut PIN) -> nb::Result<u16, Self::Error> {
            self.read_raw_internal(T::CHANNEL).map_err(to_nb_err)
        }
    }
```
I'm assuming that we want to pass in the AdcChannelDriver as the "pin" - but 
```
error[E0207]: the type parameter `M` is not constrained by the impl trait, self type, or predicates
   --> src/adc.rs:750:17
    |
750 |     impl<'d, T, M, PIN> embedded_hal_0_2::adc::OneShot<T::Adc, u16, T> for AdcDriver<'d, T::Adc>
    |                 ^ unconstrained type parameter

error[E0207]: the type parameter `PIN` is not constrained by the impl trait, self type, or predicates
   --> src/adc.rs:750:20
    |
750 |     impl<'d, T, M, PIN> embedded_hal_0_2::adc::OneShot<T::Adc, u16, T> for AdcDriver<'d, T::Adc>
    |                    ^^^ unconstrained type parameter
```

Maybe I'm way off base.

Reply:
You can use the e-hal 0.2 impl for the 4.x one-shot driver as a reference though it won't be exactly 1:1. First and foremost, you need to impl the `Channel` trait on `AdcChannelDriver`. Then - here - where you use the `PIN` generic - you need to replace it directly with `AdcChannelDriver<...>`. This simplification is not strictly necessary for the 4.x one-shot driver, as there the channel driver is a passive thing, but I might so it for simplicity there as well. Here this is necessary, as you need to have access to all methods of `AdcChannelDriver`, which are otherwise not captured by the `Channel` trait.